ci/webhook-receiver.py

```diff
@@ -1,18 +1,95 @@
 from flask import Flask, jsonify, request
+
 import requests
+from ipaddress import ip_address, ip_network
+
+from Crypto.PublicKey import RSA
+from Crypto.Cipher import AES, PKCS1_OAEP
 
+import hmac
 import json
 import sys
 import pprint
 import os
 import subprocess
+import logging
 
 EVENT_URL = "/github-webhook"
 CI_NAME="onvm"
 
 app = Flask(__name__)
 
-# returns extracted data if it is an event for a PR creation or PR comment creation
+logging.getLogger('werkzeug').setLevel(logging.ERROR)
+logging.basicConfig(filename="access_log", filemode='a',
+                    format='%(asctime)s, %(name)s %(levelname)s %(message)s',
+                    datefmt='%d-%b-%y %H:%M:%S', level=logging.INFO)
+
+def get_request_info(request_ctx):
+    return "Request details: IP: {}, User: {}, Repo: {}, ID: {}, Body: {}.".format(request_ctx['src_ip'], request_ctx['user'], request_ctx['repo'], request_ctx['id'], request_ctx['body'])
+
+def log_access_granted(request_ctx, custom_msg):
+    logging.info("Access GRANTED: {}. {}".format(custom_msg, get_request_info(request_ctx)))
+
+def log_access_denied(request_ctx, custom_msg):
+    logging.info("Access DENIED: {}. {}".format(custom_msg, get_request_info(request_ctx)))
+
+def decrypt_secret():
+    secret_file = open(webhook_config['secret-file'], "rb")
+    private_key = RSA.import_key(open(webhook_config['private-key-file']).read())
+
+    enc_session_key, nonce, tag, ciphertext = \
+       [ secret_file.read(x) for x in (private_key.size_in_bytes(), 16, 16, -1) ]
+
+    # Decrypt the session key with the private RSA key
+    cipher_rsa = PKCS1_OAEP.new(private_key)
+    session_key = cipher_rsa.decrypt(enc_session_key)
+
+    # Decrypt the data with the AES session key
+    cipher_aes = AES.new(session_key, AES.MODE_EAX, nonce)
+    data = cipher_aes.decrypt_and_verify(ciphertext, tag)
+
+    # Clear memory
+    secret_file = private_key = enc_session_key = nonce = tag = ciphertext = None
+    del secret_file
+    del private_key
+    del enc_session_key
+    del nonce
+    del tag
+    del ciphertext
+    
+    return data
+
+def verify_request_ip(request_ctx):
+    src_ip = request_ctx['src_ip']
+    print (src_ip)
+    valid_ips = requests.get('https://api.github.com/meta').json()['hooks']
+
+    for ip in valid_ips:
+        if src_ip in ip_network(ip):
+            return True
+
+    return False
+
+def verify_request_secret(request_ctx):
+    header_signature = request_ctx['X-Hub-Signature']
+    if header_signature is None:
+        return False
+
+    signature = header_signature.split('=')[1]
+    secret = decrypt_secret()
```

Comment:
Are we decrypting the secret each time?  That's going to become expensive depending on how many times the CI is triggered.  

Is it possible to calculate and store the secret digest during server init, and then to use that in this function when verifying request secrets?